fbcore/src/main/java/com/facebook/datasource/RetainingDataSourceSupplier.java

```diff
@@ -0,0 +1,140 @@
+package com.facebook.datasource;
+
+import com.facebook.common.executors.CallerThreadExecutor;
+import com.facebook.common.internal.Supplier;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+import javax.annotation.concurrent.ThreadSafe;
+
+@NotThreadSafe
+public class RetainingDataSourceSupplier<T> implements Supplier<DataSource<T>> {
+
+  private final Set<RetainingDataSource> mDataSources =
+      Collections.newSetFromMap(new WeakHashMap<RetainingDataSource, Boolean>());
+
+  private Supplier<DataSource<T>> mCurrentDataSourceSupplier = null;
+
+  @Override
+  public DataSource<T> get() {
+    RetainingDataSource dataSource = new RetainingDataSource();
+    dataSource.setSupplier(mCurrentDataSourceSupplier);
+    mDataSources.add(dataSource);
+    return dataSource;
+  }
+
+  public void setSupplier(Supplier<DataSource<T>> supplier) {
```

Comment:
Perhaps name this `replaceSupplier(...)` to make it clear that this method is meant to be called several times?